ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/jdbc/JDBCAbstractSplitFilterRule.java

```diff
@@ -127,7 +127,9 @@ public void onMatch(RelOptRuleCall call, SqlDialect dialect) {
     ArrayList<RexCall> validJdbcNode = visitor.getValidJdbcNode();
     ArrayList<RexCall> invalidJdbcNode = visitor.getInvalidJdbcNode();
 
-    assert validJdbcNode.size() != 0 && invalidJdbcNode.size() != 0;
+    if( validJdbcNode.size() == 0 || invalidJdbcNode.size() == 0) {
+      return;
+    }
```

Comment:
Did you hit this assertion error? My understanding is that the previous code guarantees that this should never happen so the assertion seems valid.

Reply:
Yes, unfortunately, I was always hitting this assertion error after I fixed the rel index in the onMatch method. I'll add a test case for this.

Reply:
If we cannot split the filter then we don't need to enter the `onMatch` method. A few lines above there seems to be a method:
```
public boolean canBeSplit() {
      return !validJdbcNode.isEmpty() && !invalidJdbcNode.isEmpty();
    }
```
with the same condition. 
I think the check should be part of `matches` method so we don't fire the rule if there is nothing to do.

Reply:
Updated `JDBCSplitFilterAboveJoinRule`.`matches` to call `JDBCAbstractSplitFilterRule`.`matches`, which checks the above condition.